R/chat-parallel.R

```diff
@@ -1,3 +1,88 @@
+chat_parallel <- function(chat, prompts, max_active = 10, rpm = 500) {
+  my_parallel_responses <- function(conversations) {
+    parallel_responses(
+      provider = chat$get_provider(),
+      conversations = conversations,
+      tools = chat$get_tools(),
+      max_active = max_active,
+      rpm = rpm
+    )
+  }
+
+  # First build up list of cumulative
+  user_turns <- as_user_turns(prompts)
+  existing <- chat$get_turns(include_system_prompt = TRUE)
+  conversations <- append_turns(list(existing), user_turns)
+
+  # Now get the assistants response
+  assistant_turns <- my_parallel_responses(conversations)
+  conversations <- append_turns(conversations, assistant_turns)
+
+  repeat {
+    assistant_turns <- map(
+      assistant_turns,
+      \(turn) match_tools(turn, tools = chat$get_tools())
+    )
+    user_turns <- lapply(assistant_turns, invoke_tools)
+    needs_iter <- !map_lgl(user_turns, is.null)
+    if (!any(needs_iter)) {
+      break
+    }
+
+    # don't need to index because user_turns null
+    conversations <- append_turns(conversations, user_turns)
+
+    assistant_turns <- ifelse(
+      needs_iter,
+      my_parallel_responses(conversations[needs_iter]),
+      list(NULL)
+    )
```

Comment:
The `ifelse()` here is the problem, it's only using the first element in the `yes` list.

``` r
ifelse(c(TRUE, FALSE, TRUE, FALSE), list(list(1), list(2)), list(NULL))
#> [[1]]
#> [[1]][[1]]
#> [1] 1
#> 
#> 
#> [[2]]
#> NULL
#> 
#> [[3]]
#> [[3]][[1]]
#> [1] 1
#> 
#> 
#> [[4]]
#> NULL
```

Here's a fix that I've confirmed with @simonpcouch's reprex.

```suggestion
    assistant_turns <- vector("list", length(user_turns))
    assistant_turns[needs_iter] <- my_parallel_responses(
      conversations[needs_iter]
    )
```

Reply:
With the fix

``` r
pkgload::load_all()
#> ℹ Loading ellmer

chat <- chat_openai("Be terse.", model = "gpt-4.1-nano")
rng <- local({
  count <- 0
  function() {
    count <<- count + 1
    count
  }
})

chat$register_tool(tool(rng, "Rolls a six-sided die."))

prompts <- c(
  "Roll the dice, please! Reply with 'You rolled ____'.",
  "reply with the word 'boop'",
  "Roll the dice, please! Reply with 'You rolled ____'.",
  "reply with the word 'beep'"
)
res <- ellmer:::chat_parallel(chat, as.list(prompts))

for (i in seq_along(res)) {
  cli::cli_inform("Chat {i}: {res[[i]]$last_turn()@text}")
}
#> Chat 1: You rolled 1
#> Chat 2: boop
#> Chat 3: You rolled 2.
#> Chat 4: beep
```

Without the fix

```r
for (i in seq_along(res)) {
  cli::cli_inform("Chat {i}: {res[[i]]$last_turn()@text}")
}
#> Chat 1: You rolled 1
#> Chat 2: boop
#> Chat 3: You rolled 1
#> Chat 4: beep
```